quinn/src/work_limiter.rs

```diff
@@ -0,0 +1,235 @@
+use std::time::{Duration, Instant};
+
+/// Limits the amount of time spent on a certain type of work in a cycle
+///
+/// The limiter works dynamically: For a sampled subset of cycles it measures the
+/// the time that is approximately required for fulfilling 1 work item, and
+/// calculates the amount of allowed work items per cycle.
+/// The estimates are smoothed over all cycles where the exact duration is measured.
+///
+/// In cycles where no measurement is performed the previously determined work limit
+/// is used.
+///
+/// For the limiter the exact definition of a work item does not matter.
+/// It could for example track the amount of transmitted bytes per cycle,
+/// or the amount of transmitted datagrams per cycle.
+/// It will however work best if the required time to complete a work item is
+/// constant.
+#[derive(Debug)]
+pub struct WorkLimiter {
+    /// Whether to measure the required work time, or to use the previous estimates
+    mode: LimiterMode,
+    /// The current cycle number
+    cycle: u16,
+    /// The time the cycle started - only used in measurement mode
+    start_time: Instant,
+    /// How many work items have been completed in the cycle
+    completed_work_items: usize,
+    /// The amount of work items which are allowed for a cycle
+    allowed_work_items: usize,
+    /// The desired cycle time
+    desired_cycle_time: Duration,
+    /// The estimated and smoothed time per work item in nanoseconds
+    smoothed_time_per_work_item_nanos: f64,
+}
+
+impl WorkLimiter {
+    pub fn new(desired_cycle_time: Duration) -> Self {
+        Self {
+            mode: LimiterMode::Measure,
+            cycle: 0,
+            start_time: Instant::now(),
+            completed_work_items: 0,
+            allowed_work_items: 0,
+            desired_cycle_time,
+            smoothed_time_per_work_item_nanos: 0.0,
+        }
+    }
+
+    /// Starts one work cycle
+    pub fn start_cycle(&mut self) {
+        self.completed_work_items = 0;
+        match self.mode {
+            LimiterMode::Measure => {
+                self.start_time = Instant::now();
+            }
+            LimiterMode::HistoricData => {}
+        }
+    }
+
+    /// Returns whether more work can be performed inside the `desired_cycle_time`
+    ///
+    /// Requires that previous work was tracked using `record_work`.
+    pub fn allow_work(&mut self) -> bool {
+        match self.mode {
+            LimiterMode::Measure => self.start_time.elapsed() < self.desired_cycle_time,
+            LimiterMode::HistoricData => self.completed_work_items < self.allowed_work_items,
+        }
+    }
```

Comment:
I feel like we should maybe leverage the type system more here? For example, `start() -> Cycle`, `Cycle::continue() -> bool`, `Cycle::record()`, and `impl Drop for Cycle`.

Reply:
The idea is neat. But I have been bitten so many times before where it didn't work in the end, because another call to `self.something()` was required inside the loop and the borrow-checker didn't allow it. Therefore I didn't try here.

While it seems to work with the current code (nothing inside `drive_send/drive_recv` makes a `self.` call - it might be an annoyance down the road. If this is accepted and preferred I can change it.

The reason this works better with `Mutex/RefCell` is that they are often the only field within a struct, so there is less desire to make additional `self` calls besides locking the mutex.

Reply:
Personally I would prefer it; we can tackle any borrowing issues when they arise. But let's see what @Ralith thinks.

Reply:
I don't have a super strong opinion; APIs that are constructively impossible to abuse are nice, but this case doesn't strike me as incredibly error prone to begin with, and keeping the impl simple is nice too.

Reply:
Okay, let's keep it.